### PR TITLE
[DON'T MERGE, PREVIEW ONLY] Set up spatial entities

### PR DESCRIPTION
**Pull request by drummyfish**

### CODE REVIEW

File: components/spatial_entity/loader.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#define SPATIAL_ENTITY_LOADER_MODULE_STR "spatial entity loader"
+
+namespace MFFormat
+{
+
+/**
+  Handles creating spatial entities from different loaded files.
+ */
+
+class SpatialEntityLoaderImplementation
```
Comment (drummyfish) [inline components/spatial_entity/loader.hpp]: Ideas for a better/shorter name? Should there be a super class for this?

File: components/physics/bullet_physics_world.hpp [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        return cb.mResult->getUserIndex();
+
+    return -1;
+}
+
+void BulletPhysicsWorld::frame(double dt)
+{
+    mWorld->stepSimulation(dt, 1);
+
+    size_t numManifolds = mWorld->getDispatcher()->getNumManifolds();
```
Comment (zaklaus) [inline components/physics/bullet_physics_world.hpp]: I will get rid of this part, as it serves no purpose right now.

Comment (drummyfish) [reply components/physics/bullet_physics_world.hpp]: ok

File: components/spatial_entity/loader.hpp (hunk 1 above)
Comment (zaklaus) [reply components/spatial_entity/loader.hpp]: Possibly introduce EntityLoader super class for (renamed) SpatialEntityLoader.

File: components/spatial_entity/loader.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+                // find the corresponding collision:
+               
+                for (int i = 0; i < mTreeKlzBodies->size(); ++i)    // FIXME: ugly and slow?
+                { 
+                    if ((*mTreeKlzBodies)[i].mName.compare(newEntity.getName()) == 0)
+                    {
+                        newEntity.setBulletBody((*mTreeKlzBodies)[i].mRigidBody.mBody.get());
+                        mTreeKlzBodies->erase(mTreeKlzBodies->begin() + i);
+                        break;    // TODO: can a node has multiple collisions/the other way around?
```
Comment (zaklaus) [inline components/spatial_entity/loader.hpp]: I believe dealing with up to 1 collision per static mesh (granted, tree.klz links each collision once per mesh), we probably can `break` it there. If, for some reason, tree.klz links multiple collisions to 1 particular mesh, that changes however.

File: components/spatial_entity/spatial_entity.hpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    struct Vec3s operator*(float v) { struct Vec3s r; r.x = x * v, r.y = y * v; r.z = z * v; return r; };
+    std::string str() { return "[" + std::to_string(x) + ", " + std::to_string(y) + ", " + std::to_string(z) + "]"; };
+} Vec3;
+
+typedef struct
+{
+    float x;
+    float y;
+    float z;
+} Quat;
```
Comment (zaklaus) [inline components/spatial_entity/spatial_entity.hpp]: Missing W field in this structure.

File: components/spatial_entity/spatial_entity.hpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    bool mReady;
+    int mID;
+};
+
+int SpatialEntity::sNextID = 0;
+
+SpatialEntity::SpatialEntity()
+{
+    mID = sNextID;
+    sNextID++;
```
Comment (zaklaus) [inline components/spatial_entity/spatial_entity.hpp]: Discussion: I could provide the necessary ID allocation, with the support of entity removal as well.

File: components/spatial_entity/spatial_entity_implementation.hpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    virtual std::string toString() override;
+
+    void setOSGNode(osg::MatrixTransform *t)  { mOSGNode = t;          };
+    void setBulletBody(btRigidBody *body)     { mBulletBody = body;    };
+    osg::MatrixTransform *getOSGNode()        { return mOSGNode.get(); };
+    btRigidBody *getBulletBody()              { return mBulletBody;    };
+    void setOSGRootNode(osg::Group *root)     { mOSGRoot = root;       };
+
+protected:
+    void makePhysicsDebugOSGNode();        ///< Creates a visual representation of the physical representation.
```
Comment (zaklaus) [inline components/spatial_entity/spatial_entity_implementation.hpp]: Possibly move this to new Debug class or somehow decouple debug code from here.

**Pull request merged**